File: core/utils/port.cpp
#include <utils/port.h>

Port8Bit::Port8Bit(uint16_t port){
	portnumber = port;
	offset = 0;
}

Port8Bit::~Port8Bit(){
	
}

uint8_t Port8Bit::Read(){
	uint8_t result;
	__asm__ ("inb %1, %0" : "=a" (result) : "Nd" ((uint16_t) portnumber + offset));
	return result;
}

void Port8Bit::Write(uint8_t data){
	__asm__ ("outb %0, %1" : : "a" (data), "Nd" ((uint16_t) portnumber + offset));
}

Port16Bit::Port16Bit(uint16_t port){
	portnumber = port;
	offset = 0;
}

Port16Bit::~Port16Bit(){
	
}

uint16_t Port16Bit::Read(){
	uint16_t result;
	__asm__ volatile("inw %1, %0" : "=a" (result) : "Nd" ((uint16_t) portnumber + offset));
	return result;
}

void Port16Bit::Write(uint16_t data){
	asm volatile ("outw %0, %1" : : "a" (data), "Nd" ((uint16_t) portnumber + offset));
}

Port32Bit::Port32Bit(uint16_t port){
	portnumber = port;
	offset = 0;
}

Port32Bit::~Port32Bit(){
	
}

uint32_t Port32Bit::Read(){
	uint32_t result;
	asm volatile ("inl %1, %0" : "=a" (result) : "d" ((uint16_t) portnumber + offset));
	return result;
}

void Port32Bit::Write(uint32_t data){
	asm volatile ("outl %0, %1" : : "a" (data), "Nd" ((uint16_t) portnumber + offset));
}

void io_wait() {
	asm volatile ("outb %%al, $0x80" : : "a"(0));
}

unsigned char inb(unsigned short _port){
	unsigned char result;
	__asm__ ("inb %1, %0" : "=a" (result) : "Nd" (_port));
	return result;
}

void outb(unsigned short _port, unsigned char _data){
	__asm__ ("outb %0, %1" : : "a" (_data), "Nd" (_port));
}

uint16_t inw(uint16_t _port) {
	uint16_t result;
	__asm__ volatile("inw %1, %0" : "=a" (result) : "Nd" (_port));
	return result;
}

void outw(uint16_t _port, uint16_t data) {
	asm volatile ("outw %0, %1" : : "a" (data), "Nd" (_port));
}

uint32_t inl(uint16_t _port) {
	uint32_t result;
	__asm__ volatile("inl %1, %0" : "=a" (result) : "Nd" (_port));
	return result;
}

void outl(uint16_t _port, uint32_t data) {
	asm volatile ("outl %0, %1" : : "a" (data), "Nd" (_port));
}

File: foxos_rs/src/port.rs
extern {
	fn __rust_io_wait();

	fn __rust_inb(port: u16) -> u8;
	fn __rust_outb(port: u16, data: u8);
	fn __rust_inw(port: u16) -> u16;
	fn __rust_outw(port: u16, data: u16);
	fn __rust_inl(port: u16) -> u32;
	fn __rust_outl(port: u16, data: u32); 
}

pub fn io_wait() {
	unsafe {
		__rust_io_wait()
	}
}

pub fn inb(port: u16) -> u8 {
	unsafe {
		__rust_inb(port)
	}
}

pub fn outb(port: u16, data: u8) {
	unsafe {
		__rust_outb(port, data)
	}
}

pub fn inw(port: u16) -> u16 {
	unsafe {
		__rust_inw(port)
	}
}

pub fn outw(port: u16, data: u16) {
	unsafe {
		__rust_outw(port, data)
	}
}

pub fn inl(port: u16) -> u32 {
	unsafe {
		__rust_inl(port)
	}
}

pub fn outl(port: u16, data: u32) {
	unsafe {
		__rust_outl(port, data)
	}
}